modules/substrate/src/lib.rs

```diff
@@ -31,43 +31,95 @@
 // Runtime-generated enums
 #![allow(clippy::large_enum_variant)]
 
-use bp_substrate::{AuthoritySet, ImportedHeader, ScheduledChange};
+use crate::storage::{AuthoritySet, ImportedHeader, ScheduledChange};
+use codec::{Codec, EncodeLike};
 use frame_support::{decl_error, decl_module, decl_storage, dispatch};
 use frame_system::ensure_signed;
-use sp_runtime::traits::Header as HeaderT;
-use sp_std::{marker::PhantomData, prelude::*};
+use num_traits::AsPrimitive;
+use sp_runtime::traits::{
+	AtLeast32BitUnsigned, Hash as HashT, Header as HeaderT, MaybeDisplay, MaybeMallocSizeOf, MaybeSerializeDeserialize,
+	Member, SimpleBitOps,
+};
+use sp_std::{fmt::Debug, marker::PhantomData, prelude::*, str::FromStr};
 
 mod justification;
+mod storage;
 mod storage_proof;
 mod verifier;
 
 #[cfg(test)]
 mod mock;
 
-type Hash<T> = <T as HeaderT>::Hash;
-type Number<T> = <T as HeaderT>::Number;
-
-pub trait Trait: frame_system::Trait {}
+pub trait Trait: frame_system::Trait {
+	/// A type that fulfulls the abstract idea of what a Substrate header is.
+	// See here for more info:
+	// https://crates.parity.io/sp_runtime/traits/trait.Header.html
+	type BridgedHeader: HeaderT<Number = Self::BridgedBlockNumber, Hash = Self::BridgedBlockHash>;
+
+	/// A type that fulfills the abstract idea of what a Substrate block number is.
+	// Constraits come from the associated Number type of `sp_runtime::traits::Header`
+	// See here for more info:
+	// https://crates.parity.io/sp_runtime/traits/trait.Header.html#associatedtype.Number
+	//
+	// Note that the `AsPrimitive<usize>` trait is required by the Grandpa justification
```

Comment:
Why is it `usize`? We shouldn't be using that in the runtime, should be at least `u64`.

Reply:
I didn't make the rules, I just follow them (trait definition [here](https://github.com/paritytech/finality-grandpa/blob/master/src/lib.rs#L169))